osu.Game/Beatmaps/BeatmapManager_UpdateQueue.cs

```diff
@@ -0,0 +1,180 @@
+
+                if (api?.State != APIState.Online)
+                    return;
+
+                var req = new GetBeatmapRequest(beatmap);
+
+                req.Failure += fail;
+
+                try
+                {
+                    // intentionally blocking to limit web request concurrency
+                    api.Perform(req);
+
+                    var res = req.Result;
+
+                    if (res != null)
+                    {
+                        beatmap.Status = res.Status;
+                        beatmap.BeatmapSet.Status = res.BeatmapSet.Status;
+                        beatmap.BeatmapSet.OnlineBeatmapSetID = res.OnlineBeatmapSetID;
+                        beatmap.OnlineBeatmapID = res.OnlineBeatmapID;
+
+                        LogForModel(set, $"Online retrieval mapped {beatmap} to {res.OnlineBeatmapSetID} / {res.OnlineBeatmapID}.");
+                    }
+                }
+                catch (Exception e)
+                {
+                    fail(e);
+                }
+
+                void fail(Exception e)
+                {
+                    beatmap.OnlineBeatmapID = null;
+                    LogForModel(set, $"Online retrieval failed for {beatmap} ({e.Message})");
+                }
+            }
+
+            private void prepareLocalCache()
+            {
+                string cacheFilePath = storage.GetFullPath(cache_database_name);
+                string compressedCacheFilePath = $"{cacheFilePath}.bz2";
+
+                cacheDownloadRequest = new FileWebRequest(compressedCacheFilePath, $"https://assets.ppy.sh/client-resources/{cache_database_name}.bz2");
+
+                cacheDownloadRequest.Failed += ex =>
+                {
+                    File.Delete(compressedCacheFilePath);
+                    File.Delete(cacheFilePath);
+
+                    Logger.Log($"{nameof(BeatmapUpdateQueue)}'s online cache download failed: {ex}", LoggingTarget.Database);
+                };
+
+                cacheDownloadRequest.Finished += () =>
+                {
+                    try
+                    {
+                        using (var stream = File.OpenRead(cacheDownloadRequest.Filename))
+                        using (var outStream = File.OpenWrite(cacheFilePath))
+                        using (var bz2 = new BZip2Stream(stream, CompressionMode.Decompress, false))
+                            bz2.CopyTo(outStream);
+
+                        // set to null on completion to allow lookups to begin using the new source
+                        cacheDownloadRequest = null;
+                    }
+                    catch (Exception ex)
+                    {
+                        Logger.Log($"{nameof(BeatmapUpdateQueue)}'s online cache extraction failed: {ex}", LoggingTarget.Database);
+                    }
+                    finally
+                    {
+                        File.Delete(compressedCacheFilePath);
+                        File.Delete(cacheFilePath);
```

Comment:
Are you sure? I think you might be missing an extraction failure check on this particular deletion, because as it stands when I run this code it downloads the archive fine, then unzips it and promptly nukes both files right after that, rendering the cache lookup code a little pointless...